### PR TITLE
Make agent environment threadsafe

### PR DESCRIPTION
**Pull request by moskyb**
In #1943, we're adding a feature to allow a server that's running in a goroutine to modify the environment of the agent, via calls to its API. Without realising it, i created a PR that was essentially just one giant race condition - the environment wasn't guaranteed to be synchronised between any number of goroutines accessing it. This made the golang race unhappy, and also meant that the env might get out of sync in some situations.

**This PR:** Updates the `env.Environment` type's implementation to make it backed by an [`xsync.MapOf[string, string]`](https://github.com/puzpuzpuz/xsync#map), rather than it being a plain map with some sauce on it. this makes it so that both reads and writes are synchronised across goroutines, meaning that we won't get data races when accessing the env across threads.

**Wait, why didn't you just...**
- Wrap the existing calls in a mutex?
  - Doing so would've still meant converting the env.Environment from being a map to being a struct that holds a map (and a mutex), which would've been a similar amount of effort to this
- Use a `sync.Map`?
  - `sync.Map`s are untyped, and casting everything coming out of it is a pain and kind of ugly

### CODE REVIEW

File: env/environment.go [hunk 1]
```diff
@@ -70,47 +99,54 @@ func (e Environment) GetBool(key string, defaultValue bool) bool {
...
+	_, ok := e.underlying.Load(key)
 	return ok
 }
 
 // Set sets a key in the environment
-func (e Environment) Set(key string, value string) string {
-	e[normalizeKeyName(key)] = value
-
+func (e *Environment) Set(key string, value string) string {
+	e.underlying.Store(key, value)
```
Comment (triarius) [inline env/environment.go:109]: needs `normalizeKeyName`?

**Review by triarius — DISMISSED** [review]
Must have been so frustration to go through the race tests to figure this out!

I have a few questions, but I agree with the general approach.

File: env/environment.go [hunk 2]
```diff
@@ -70,47 +99,54 @@ func (e Environment) GetBool(key string, defaultValue bool) bool {
 }
 
 // Exists returns true/false depending on whether or not the key exists in the env
-func (e Environment) Exists(key string) bool {
-	_, ok := e[normalizeKeyName(key)]
+func (e *Environment) Exists(key string) bool {
+	_, ok := e.underlying.Load(key)
```
[the remark this answers is not in the document]
Comment (moskyb) [reply env/environment.go:103]: 😬 whoooops. shocked that the tests don't cover this TBH...

File: env/environment.go [hunk 3]
```diff
@@ -8,14 +8,32 @@ import (
...
+
+	"github.com/puzpuzpuz/xsync/v2"
 )
 
 // Environment is a map of environment variables, with the keys normalized
 // for case-insensitive operating systems
-type Environment map[string]string
+type Environment struct {
+	underlying *xsync.MapOf[string, string]
+}
```
[the remark this answers is not in the document]
Comment (moskyb) [reply env/environment.go:19]: potentially not? but it's what the `NewMapOf*` funcs return, and i thiiiiiiiiink(????) that we want the base struct to always be passed by reference? i'm not 100% sure on that though

[the remark this answers is not in the document]
Comment (triarius) [reply env/environment.go:19]: Hmm, so we always pass around a `*Environment` yeah? So it seems unnecessary to have it also contain a pointer.

Maybe if we change `Environment` to be an interface, but then the interface definition should be in the consuming packages, as per https://github.com/golang/go/wiki/CodeReviewComments#interfaces. That seems like a more annoying refactor. So this looks good to me.

File: env/environment.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (triarius) [reply env/environment.go:103]: > shocked

File: env/environment_test.go [hunk 4]
```diff
@@ -31,6 +33,48 @@ func TestEnvironmentSet(t *testing.T) {
 	assert.True(t, ok)
 }
 
+func TestEnvironmentSet_NormalizesKeyNames(t *testing.T) {
```
Comment (triarius) [inline env/environment_test.go:36]: Need to test `Exists` as well

File: env/environment.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (triarius) [reply env/environment.go:103]: Exist is still missing `normalizeKeyName` as of [`b5459b2` (#2013)](https://github.com/buildkite/agent/pull/2013/commits/b5459b21bd169433d932d9e22e51d8afe56ff643)

**Review by triarius — APPROVED** [review]
LGTM

**Pull request merged**